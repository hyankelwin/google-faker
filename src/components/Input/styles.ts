import styled, { css } from 'styled-components';
import Color from '../../styles/color';

interface ContainerProps {
  isMedium: boolean;
}

export const Container = styled.div<ContainerProps>`
  border-radius: 50px;
  border: 1px solid ${Color.grey[200]};
  padding: 16px;
  width: 100%;
  margin-top: 10px;
  margin-bottom: 10px;
  color: #666360;

  ${props =>
    props.isMedium &&
    css`
      width: 400px;
      height: 30px;

      @media (max-width: 700px) {
        width: 300px;
      }

      @media (max-width: 440px) {
        width: 250px;
      }
    `}

  display: flex;
  align-items: center;

  & + div {
    margin-top: 8px;
  }

  :hover {
    box-shadow: 2px 2px 8px rgba(0, 0, 0, 0.04);
  }

  input {
    flex: 1;
    background: transparent;
    border: 0;
    color: #1c1c1c;
    width: 500px;
    &::placeholder {
      color: #666360;
    }

    @media (max-width: 600px) {
      width: 300px;
    }

    @media (max-width: 390px) {
      width: 200px;
    }
  }

  svg {
    margin-right: 16px;
    opacity: 0.5;

    :nth-child(3) {
      margin-right: 1px;
    }
  }
`;
