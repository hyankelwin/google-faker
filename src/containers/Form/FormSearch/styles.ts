import styled, { css } from 'styled-components';
import Color from '../../../styles/color';

interface ContainerProps {
  isMedium: boolean;
}

export const ContentForm = styled.div<ContainerProps>`
  background-color: ${Color.white};
  padding: 20px 40px;
  border-radius: 8px;

  ${props =>
    props.isMedium &&
    css`
      padding: 0px 0px;
    `}
`;

export const ContentButton = styled.div`
  width: 100px;
  margin: auto;
  padding-top: 10px;
`;
