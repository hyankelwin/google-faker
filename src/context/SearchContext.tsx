import React, { createContext, useState, useCallback, useContext } from 'react';
import { Animal } from '../api/google-faker-api/search/interfaces/animal.interface';
import { getAnimals } from '../api/google-faker-api/search/google-faker.request';
import { getTimeOut } from '../utils/getTimeOutSimulateApi';
import { checkAnimals } from '../utils/getAnimalType';

interface ISearch {
  isLoading: boolean;
  error: boolean;
  search: string;
  animals: Animal[];
  animalSelected: Animal;
  showCard: boolean;
  visibilityModalOptions: boolean;
  visibilityModalSearch: boolean;
  inputIsEmpty: boolean;
  handleFindSeach(value: string): void;
  handleClearInputValue(): void;
  handleSetItemSelected(item: Animal): void;
  handleSetShowCard(): void;
  handleSetClearCard(): void;
  handleSetVisibiltyModalOptions(value: boolean): void;
  handleSetVisibiltyModalSearch(value: boolean): void;
  handleSetEmptyInputValue(value: boolean): void;
}

const SearchContext = createContext<ISearch>({} as ISearch);

const SearchProvider: React.FC = ({ children }) => {
  const [search, setSearch] = useState('');
  const [showCard, setShowCard] = useState(false);
  const [inputIsEmpty, setInputIsEmpty] = useState(true);
  const [visibilityModalOptions, setVisibilityModalOptions] = useState(false);
  const [visibilityModalSearch, setVisibilityModalSearch] = useState(false);
  const [isLoading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  const [animals, setAnimals] = useState<Animal[]>([]);
  const [animalSelected, setAnimalSelected] = useState<Animal>({} as Animal);

  const handleSetItemSelected = useCallback((item: Animal) => {
    setAnimalSelected(item);
  }, []);

  const handleClearInputValue = useCallback(() => {
    setSearch('');
  }, []);

  const handleSetEmptyInputValue = useCallback(value => {
    setInputIsEmpty(value);
  }, []);

  const handleSetShowCard = useCallback(() => {
    setShowCard(true);
  }, []);

  const handleSetClearCard = useCallback(() => {
    setShowCard(false);
  }, []);

  const handleSetLocalStorage = useCallback((value: string) => {
    setShowCard(false);
    localStorage.setItem('search', value);
  }, []);

  const handleSetVisibiltyModalOptions = useCallback((value: boolean) => {
    setVisibilityModalOptions(value);
  }, []);

  const handleSetVisibiltyModalSearch = useCallback((value: boolean) => {
    setVisibilityModalSearch(value);
  }, []);

  const handleFindSeach = useCallback(
    async (value: string) => {
      setLoading(true);
      setError(false);
      setSearch(value);
      handleSetLocalStorage(value);

      await getTimeOut(10);

      if (!checkAnimals(value)) {
        setLoading(false);
        setError(true);
        return;
      }

      const response = await getAnimals(value);

      setAnimals(response);
      setLoading(false);
    },
    [handleSetLocalStorage],
  );

  return (
    <SearchContext.Provider
      value={{
        search,
        isLoading,
        error,
        animals,
        animalSelected,
        showCard,
        visibilityModalOptions,
        visibilityModalSearch,
        inputIsEmpty,
        handleFindSeach,
        handleClearInputValue,
        handleSetItemSelected,
        handleSetShowCard,
        handleSetClearCard,
        handleSetVisibiltyModalOptions,
        handleSetVisibiltyModalSearch,
        handleSetEmptyInputValue,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

function useSearch(): ISearch {
  const context = useContext(SearchContext);

  return context;
}

export { SearchProvider, useSearch };
