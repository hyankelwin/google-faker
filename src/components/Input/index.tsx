import React, { InputHTMLAttributes, useEffect, useRef } from 'react';
import { IconBaseProps } from 'react-icons';
import { FiX } from 'react-icons/fi';
import { useField } from '@unform/core';
import { Container } from './styles';

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  name: string;
  icon?: React.ComponentType<IconBaseProps>;
  isMedium?: boolean;
  isIconClose?: boolean;
  onClearInput: () => void;
  isEmpty?: boolean;
}

const Input: React.FC<InputProps> = ({
  name,
  icon: Icon,
  isMedium = false,
  isIconClose = false,
  onClearInput,
  isEmpty = true,
  ...rest
}) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const { fieldName, defaultValue, registerField } = useField(name);

  useEffect(() => {
    registerField({
      name: fieldName,
      ref: inputRef.current,
      path: 'value',
    });
  }, [fieldName, registerField]);

  return (
    <Container isMedium={isMedium}>
      {Icon && <Icon size={20} />}
      <input
        data-testid="input-container"
        defaultValue={defaultValue}
        ref={inputRef}
        {...rest}
      />
      {isIconClose && !isEmpty && (
        <FiX data-testid="icon-fix" onClick={() => onClearInput()} size={20} />
      )}
    </Container>
  );
};

export default Input;
