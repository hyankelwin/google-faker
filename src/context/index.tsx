import React from 'react';

import { SearchProvider } from './SearchContext';

const AppProvider: React.FC = ({ children }) => (
  <SearchProvider>{children}</SearchProvider>
);

export default AppProvider;
