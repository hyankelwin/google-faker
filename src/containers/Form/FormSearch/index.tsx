import React, { useRef, useCallback } from 'react';
import { FiSearch } from 'react-icons/fi';
import { Form } from '@unform/web';
import { FormHandles } from '@unform/core';
import { useHistory } from 'react-router-dom';
import { ContentForm, ContentButton } from './styles';
import Input from '../../../components/Input';
import Button from '../../../components/Button';
import Color from '../../../styles/color';
import { useSearch } from '../../../context/SearchContext';
import Paths from '../../../constants/paths.constant';

interface FormSearchProps {
  showButton?: boolean;
  isMedium?: boolean;
  page?: string;
}

const FormSearch: React.FC<FormSearchProps> = ({
  showButton = true,
  isMedium = false,
  page,
}) => {
  const history = useHistory();
  const { handleFindSeach, search, handleSetEmptyInputValue, inputIsEmpty } =
    useSearch();

  const formRef = useRef<FormHandles>(null);

  const handleSubmit = useCallback(
    async values => {
      handleFindSeach(values.search);
      if (page === 'home') {
        history.push(Paths.Search);
      }
    },
    [handleFindSeach, history, page],
  );

  const handleClearInput = useCallback(() => {
    handleSetEmptyInputValue(true);
    formRef.current?.reset();
  }, [handleSetEmptyInputValue]);

  return (
    <Form ref={formRef} onSubmit={handleSubmit}>
      <ContentForm isMedium={isMedium}>
        <Input
          name="search"
          icon={FiSearch}
          data-testid="input-search"
          onChange={data => handleSetEmptyInputValue(!data.target.value)}
          isMedium={isMedium}
          defaultValue={page === 'search' ? search : ''}
          isIconClose={page === 'search'}
          onClearInput={() => handleClearInput()}
          isEmpty={inputIsEmpty}
        />
        {showButton && (
          <ContentButton>
            <Button
              type="submit"
              data-testid="button-submit"
              color={Color.black.text}
              background={Color.grey[200]}
              disabled={inputIsEmpty}
            >
              Buscar
            </Button>
          </ContentButton>
        )}
      </ContentForm>
    </Form>
  );
};

export default FormSearch;
