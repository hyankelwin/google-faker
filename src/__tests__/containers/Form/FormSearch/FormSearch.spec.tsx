import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { useSearch } from '../../../../context/SearchContext';
import FormSearch from '../../../../containers/Form/FormSearch';
import Input from '../../../../components/Input';

const mockedHistoryPush = jest.fn();

jest.mock('react-router-dom', () => {
  return {
    useHistory: () => ({
      push: mockedHistoryPush,
    }),
    Link: ({ children }: { children: React.ReactNode }) => children,
  };
});

jest.mock('../../../../context/SearchContext');

describe('FormSearch container', () => {
  it('should be able to find on FormSearch container with success', async () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleFindSeach: () => jest.fn(),
      handleSetEmptyInputValue: () => jest.fn(),

      search: 'dog',
    });

    const { getByTestId, asFragment } = render(<FormSearch />);

    const searchField = getByTestId('input-search');
    const buttonElement = getByTestId('button-submit');

    fireEvent.change(searchField, { target: { value: 'dog' } });

    fireEvent.click(buttonElement);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should be able to find on FormSearch container when page is home', async () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleFindSeach: () => jest.fn(),
      handleSetEmptyInputValue: () => jest.fn(),
      search: 'dog',
    });

    const props = {
      page: 'home',
    };

    const { getByTestId, asFragment } = render(<FormSearch {...props} />);

    const searchField = getByTestId('input-search');
    const buttonElement = getByTestId('button-submit');

    fireEvent.change(searchField, { target: { value: 'dog' } });

    fireEvent.click(buttonElement);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should be able to simulate click on FormSearch container', async () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleFindSeach: () => jest.fn(),
      handleSetEmptyInputValue: () => jest.fn(),
      search: '',
      inputIsEmpty: false,
    });

    const props = {
      page: 'search',
      onClearInput: jest.fn(),
    };

    const propsInput = {
      name: 'search',
      isEmpty: false,
      onClearInput: jest.fn(),
      isIconClose: true,
    };
    const { getByTestId, asFragment } = render(
      <FormSearch {...props}>
        <Input {...propsInput} />
      </FormSearch>,
    );

    const searchField = getByTestId('input-search');
    const iconClearField = getByTestId('icon-fix');

    fireEvent.change(searchField, { target: { value: 'dog' } });

    fireEvent.click(iconClearField);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
