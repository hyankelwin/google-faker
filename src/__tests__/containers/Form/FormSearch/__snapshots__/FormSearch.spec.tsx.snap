// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSearch container should be able to find on FormSearch container when page is home 1`] = `
<DocumentFragment>
  <form>
    <div
      class="sc-aXZVg jGGoBE"
    >
      <div
        class="sc-eqUAAy kzGVwk"
      >
        <svg
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <input
          data-testid="input-search"
          value=""
        />
      </div>
      <div
        class="sc-gEvEer gYDKcr"
      >
        <button
          class="sc-fqkvVR fJaxrU"
          color="rgba(0,0,0,0.87)"
          data-testid="button-submit"
          type="submit"
        >
          Buscar
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`FormSearch container should be able to find on FormSearch container with success 1`] = `
<DocumentFragment>
  <form>
    <div
      class="sc-aXZVg jGGoBE"
    >
      <div
        class="sc-eqUAAy kzGVwk"
      >
        <svg
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <input
          data-testid="input-search"
          value=""
        />
      </div>
      <div
        class="sc-gEvEer gYDKcr"
      >
        <button
          class="sc-fqkvVR fJaxrU"
          color="rgba(0,0,0,0.87)"
          data-testid="button-submit"
          type="submit"
        >
          Buscar
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`FormSearch container should be able to simulate click on FormSearch container 1`] = `
<DocumentFragment>
  <form>
    <div
      class="sc-aXZVg jGGoBE"
    >
      <div
        class="sc-eqUAAy kzGVwk"
      >
        <svg
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <input
          data-testid="input-search"
          value=""
        />
        <svg
          data-testid="icon-fix"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </div>
      <div
        class="sc-gEvEer gYDKcr"
      >
        <button
          class="sc-fqkvVR fJaxrU"
          color="rgba(0,0,0,0.87)"
          data-testid="button-submit"
          type="submit"
        >
          Buscar
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
